lib/workspace.js

```diff
@@ -91,6 +60,60 @@ module.exports = {
         }
     },
 
+    /**
+     * Recursively uploads a directory to a path in S3, including empty subfolders.
+     * @param {string} s3Path Local directory to upload.
+     * @param {string} bucketName Bucket name to upload into.  Can specify a path with <bucketName>/directory.
+     * @param {array | optional} ignoreList List of files to ignore.  This should be paths relative to the base directory.
+     */
+    async uploadDirectoryToS3Async(s3Path, bucketName, ignoreList) {
+        const s3 = new AWS.S3();
+        if (typeof ignoreList === 'undefined') {
+            ignoreList = [];
+        }
+        ignoreList = new Set(ignoreList);
+
+        async function walkDirectory(directory) {
+            const files = await fsPromises.readdir(directory);
+            await async.each(files, async (file) => {
+                const filePath = path.join(directory, file);
+                const stat = await fsPromises.stat(filePath);
+                /* Get the path relative to the start of our base directory */
+                const bucketPath = filePath.substring(s3Path.length + 1);
+
+                if (stat.isFile()) {
+                    if (!ignoreList.has(bucketPath)) {
+                        const fileBody = await fsPromises.readFile(filePath);
+                        try {
+                            await s3.putObject({
+                                Bucket: bucketName,
+                                Key: bucketPath,
+                                Body: fileBody,
+                            }).promise();
+                            logger.info(`synced ${bucketPath} to ${bucketName}`);
+                        } catch (err) {
+                            logger.error(`error syncing ${bucketPath}:\n ${err}`);
+                        }
+                    }
+                } else if (stat.isDirectory()) {
+                    try {
+                        await s3.putObject({
+                            Bucket: bucketName,
+                            Key: bucketPath + '/', /* Put a forward slash at the end to create a directory */
```

Comment:
```suggestion
                            Key: bucketPath + (bucketPath.endsWith('/') ? '' : '/'), /* Put a forward slash at the end to create a directory */
```
Not sure if this is the cleanest way to do this, but I think Matt was concerned about appending an extra `'/'` if there was already a slash.